Ensure session is removed from network process before deleting data store files

szewai opened the pull request:
#### ac589568e9ca1c0ffc1542789647f627dd26c132
<pre>
Ensure session is removed from network process before deleting data store files
<a href="https://bugs.webkit.org/show_bug.cgi?id=290561">https://bugs.webkit.org/show_bug.cgi?id=290561</a>
<a href="https://rdar.apple.com/143604999">rdar://143604999</a>

Reviewed by Chris Dumez.

In existing implementation, WebKit directly deletes directory of custom data store if the corresponding WebsiteDataStore
object does not exist. However, even though WebsiteDataStore object does not exist in UI process, there is still a small
chance that network process still uses the data store path (since UI process tells network process to remove session in
the destructor of WebsiteDataStore and network process may not proceed the message on time). To avoid removing directory
when it is in use, now UI process tells network process to check whether the directory is still in use and log error if
it is, and UI process will delete the directory after receiving reply.

API Test: WKWebsiteDataStore.RemoveSessionWithIdentifierFromNetworkProcess

* Source/WebKit/NetworkProcess/NetworkProcess.cpp:
(WebKit::NetworkProcess::destroySession):
(WebKit::NetworkProcess::ensureSessionWithDataStoreIdentifierRemoved):
* Source/WebKit/NetworkProcess/NetworkProcess.h:
* Source/WebKit/NetworkProcess/NetworkProcess.messages.in:
* Source/WebKit/NetworkProcess/NetworkSession.cpp:
(WebKit::m_dataStoreIdentifier):
(WebKit::m_resourceMonitorThrottlerDirectory): Deleted.
* Source/WebKit/NetworkProcess/NetworkSession.h:
(WebKit::NetworkSession::dataStoreIdentifier const):
* Source/WebKit/UIProcess/WebsiteData/Cocoa/WebsiteDataStoreCocoa.mm:
(WebKit::WebsiteDataStore::removeDataStoreWithIdentifierImpl):
(WebKit::WebsiteDataStore::removeDataStoreWithIdentifier):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebsiteDatastore.mm:
(TestWebKitAPI::(WKWebsiteDataStore, RemoveSessionWithIdentifierFromNetworkProcess)):

Canonical link: <a href="https://commits.webkit.org/292880@main">https://commits.webkit.org/292880@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e51f84e01ca2340d336369b6143d6d105db619a1

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/97303 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/16928 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/7142 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/102389 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/47831 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/99348 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/17221 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/25377 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/74137 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/31326 "Found 1 new test failure: http/tests/security/no-indexeddb-from-sandbox.html (failure)") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/100306 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/13026 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/87996 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/54475 "Found 1 new API test failure: /WPE/TestCookieManager:/webkit/WebKitCookieManager/persistent-storage (failure)") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/12782 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/5876 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/47274 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/82824 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/5958 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/104409 "Built successfully") | 
| | [❌ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/24380 "Hash e51f84e0 for PR 43134 does not build (failure)") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/17778 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/83180 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/24753 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/84119 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/82594 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/20795 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/27145 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/4823 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/17926 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/24344 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/29497 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/24166 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/27480 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/25740 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebKit/NetworkProcess/NetworkProcess.cpp:
```diff
@@ -616,6 +616,9 @@ void NetworkProcess::destroySession(PAL::SessionID sessionID, CompletionHandler<
     if (auto session = m_networkSessions.take(sessionID)) {
+        auto dataStoreIdentifier = session->dataStoreIdentifier();
+        UNUSED_PARAM(dataStoreIdentifier);
+        RELEASE_LOG(Storage, "%p - NetworkProcess::destroySession sessionID=%" PRIu64 " identifier=%s", this, sessionID.toUInt64(), dataStoreIdentifier ? dataStoreIdentifier->toString().utf8().data() : "null"_s);
```
`%" PUBLIC_LOG_STRING "` assuming there is nothing sensitive in the databaseStoreIdentifier.

cdumez commented on Source/WebKit/NetworkProcess/NetworkProcess.cpp:
```diff
@@ -631,6 +634,17 @@ void NetworkProcess::destroySession(PAL::SessionID sessionID, CompletionHandler<
 
+void NetworkProcess::ensureSessionWithDataStoreIdentifierRemoved(WTF::UUID identifier, CompletionHandler<void()>&& completionHandler)
+{
+    RELEASE_LOG(Storage, "%p - NetworkProcess::ensureSessionWithDataStoreIdentifierRemoved identifier=%s", this, identifier.toString().utf8().data());
```
ditto.

cdumez commented on Source/WebKit/NetworkProcess/NetworkProcess.cpp:
```diff
@@ -631,6 +634,17 @@ void NetworkProcess::destroySession(PAL::SessionID sessionID, CompletionHandler<
+    RELEASE_LOG(Storage, "%p - NetworkProcess::ensureSessionWithDataStoreIdentifierRemoved identifier=%s", this, identifier.toString().utf8().data());
+    for (auto& session : m_networkSessions.values()) {
+        if (session->dataStoreIdentifier() == identifier)
+            RELEASE_LOG_ERROR(Storage, "NetworkProcess::ensureSessionWithDataStoreIdentifierRemoved session still exists for identifier %s", identifier.toString().utf8().data());
```
Ditto

cdumez commented on Source/WebKit/NetworkProcess/NetworkSession.h:
```diff
@@ -302,6 +302,8 @@ class NetworkSession : public WebCore::SWServerDelegate, public CanMakeCheckedPt
     void resetResourceMonitorThrottlerForTesting();
 #endif
 
+    std::optional<WTF::UUID> dataStoreIdentifier() const { return m_dataStoreIdentifier.asOptional(); }
```
I doubt `.asOptional()` is needed.

cdumez commented on Source/WebKit/UIProcess/WebsiteData/Cocoa/WebsiteDataStoreCocoa.mm:
```diff
@@ -336,6 +323,37 @@ static String defaultWebsiteDataStoreRootDirectory()
+    ASSERT(isMainRunLoop());
+
+    auto completionHandler = [identifier, callback = WTFMove(callback)](const String& error) mutable {
+        RELEASE_LOG(Storage, "WebsiteDataStore::removeDataStoreWithIdentifier: Removal completed for identifier %s (error '%s')", identifier.toString().utf8().data(), error.isEmpty() ? "null"_s : error.utf8().data());
```
Strings need to be logged with PUBLIC_LOG_STRING if there are not sensitive and if you want them to show in sysdiagnoses.

cdumez commented on Source/WebKit/UIProcess/WebsiteData/Cocoa/WebsiteDataStoreCocoa.mm:
```diff
@@ -336,6 +323,37 @@ static String defaultWebsiteDataStoreRootDirectory()
+        RELEASE_LOG(Storage, "WebsiteDataStore::removeDataStoreWithIdentifier: Removal completed for identifier %s (error '%s')", identifier.toString().utf8().data(), error.isEmpty() ? "null"_s : error.utf8().data());
+        callback(error);
+    };
+    RELEASE_LOG(Storage, "WebsiteDataStore::removeDataStoreWithIdentifier: Removal started for identifier %s", identifier.toString().utf8().data());
```
ditto.

cdumez commented on Source/WebKit/UIProcess/WebsiteData/Cocoa/WebsiteDataStoreCocoa.mm:
```diff
@@ -336,6 +323,37 @@ static String defaultWebsiteDataStoreRootDirectory()
+    if (!identifier.isValid())
+        return completionHandler("Identifier is invalid"_s);
+
+    if (auto existingDataStore = existingDataStoreForIdentifier(identifier)) {
```
Shouldn't we use a RefPtr here?

cdumez commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebsiteDatastore.mm:
```diff
@@ -633,6 +633,43 @@ HTTPServer server([connectionCount = 0] (Connection connection) mutable {
+    RetainPtr uuid = adoptNS([[NSUUID alloc] initWithUUIDString:uuidString]);
+    RetainPtr<NSURL> generalStorageDirectory;
+    @autoreleasepool {
+        auto websiteDataStore = createWebsiteDataStoreAndPrepare(uuid.get(), @"");
```
Looks like you may be hiding a smart pointer behind `auto` here..

szewai committed:
Ensure session is removed from network process before deleting data store files
https://bugs.webkit.org/show_bug.cgi?id=290561
rdar://143604999

Reviewed by Chris Dumez.

In existing implementation, WebKit directly deletes directory of custom data store if the corresponding WebsiteDataStore
object does not exist. However, even though WebsiteDataStore object does not exist in UI process, there is still a small
chance that network process still uses the data store path (since UI process tells network process to remove session in
the destructor of WebsiteDataStore and network process may not proceed the message on time). To avoid removing directory
when it is in use, now UI process tells network process to check whether the directory is still in use and log error if
it is, and UI process will delete the directory after receiving reply.

API Test: WKWebsiteDataStore.RemoveSessionWithIdentifierFromNetworkProcess

* Source/WebKit/NetworkProcess/NetworkProcess.cpp:
(WebKit::NetworkProcess::destroySession):
(WebKit::NetworkProcess::ensureSessionWithDataStoreIdentifierRemoved):
* Source/WebKit/NetworkProcess/NetworkProcess.h:
* Source/WebKit/NetworkProcess/NetworkProcess.messages.in:
* Source/WebKit/NetworkProcess/NetworkSession.cpp:
(WebKit::m_dataStoreIdentifier):
(WebKit::m_resourceMonitorThrottlerDirectory): Deleted.
* Source/WebKit/NetworkProcess/NetworkSession.h:
(WebKit::NetworkSession::dataStoreIdentifier const):
* Source/WebKit/UIProcess/WebsiteData/Cocoa/WebsiteDataStoreCocoa.mm:
(WebKit::WebsiteDataStore::removeDataStoreWithIdentifierImpl):
(WebKit::WebsiteDataStore::removeDataStoreWithIdentifier):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebsiteDatastore.mm:
(TestWebKitAPI::(WKWebsiteDataStore, RemoveSessionWithIdentifierFromNetworkProcess)):

Canonical link: https://commits.webkit.org/292880@main
